sdk/python/feast/pyspark/launchers.py

```diff
@@ -0,0 +1,304 @@
+            process (subprocess.Popen): Pyspark driver process, spawned by the launcher.
+            output_file_uri (str): Uri to the historical feature retrieval job output file.
+        """
+        self.job_id = job_id
+        self._process = process
+        self._output_file_uri = output_file_uri
+
+    def get_id(self) -> str:
+        return self.job_id
+
+    def get_output_file_uri(self, timeout_sec: int = None):
+        with self._process as p:
+            try:
+                p.wait(timeout_sec)
+            except Exception:
+                p.kill()
+                raise SparkJobFailure("Timeout waiting for subprocess to return")
+        if self._process.returncode != 0:
+            stderr = "" if self._process.stderr is None else self._process.stderr.read()
+            stdout = "" if self._process.stdout is None else self._process.stdout.read()
+
+            raise SparkJobFailure(
+                f"Non zero return code: {self._process.returncode}. stderr: {stderr} stdout: {stdout}"
+            )
+
+
+class DataprocRetrievalJob(RetrievalJob):
+    """
+    Historical feature retrieval job result for a Dataproc cluster
+    """
+
+    def __init__(self, job_id, operation, output_file_uri):
+        """
+        This is the returned historical feature retrieval job result for DataprocClusterLauncher.
+
+        Args:
+            job_id (str): Historical feature retrieval job id.
+            operation (google.api.core.operation.Operation): A Future for the spark job result,
+                returned by the dataproc client.
+            output_file_uri (str): Uri to the historical feature retrieval job output file.
+        """
+        self.job_id = job_id
+        self._operation = operation
+        self._output_file_uri = output_file_uri
+
+    def get_id(self) -> str:
+        return self.job_id
+
+    def get_output_file_uri(self, timeout_sec=None):
+        try:
+            self._operation.result(timeout_sec)
+        except Exception as err:
+            raise SparkJobFailure(err)
+        return self._output_file_uri
+
+
+class JobLauncher(abc.ABC):
+    """
+    Submits spark jobs to a spark cluster. Currently supports only historical feature retrieval jobs.
+    """
+
+    @abc.abstractmethod
+    def historical_feature_retrieval(
+        self,
+        pyspark_script: str,
+        entity_source_conf: Dict,
+        feature_tables_sources_conf: List[Dict],
+        feature_tables_conf: List[Dict],
+        destination_conf: Dict,
+        job_id: str,
+        **kwargs,
+    ) -> RetrievalJob:
+        """
+        Submits a historical feature retrieval job to a Spark cluster.
+
+        Args:
+            pyspark_script (str): Local file path to the pyspark script for historical feature
+                retrieval.
+            entity_source_conf (List[Dict]): Entity data source configuration.
+            feature_tables_sources_conf (Dict): List of feature tables data sources configurations.
```

Comment:
Where have you documented where `feature_tables_sources_conf` will be used vs sources found in `feature_tables_conf`

Reply:
Sources would not be part of `feature_tables_conf` , as per the design of configuration found here: https://github.com/feast-dev/feast/blob/master/spark/ingestion/src/main/scala/feast/ingestion/IngestionJobConfig.scala#L74-L79

Though, i understand that the above behaviour is actually inconsistent with the way we currently define our FeatureTable proto.

Should i revamp the configuration such that `Source` is part of Feature Table conf? In which case, it would not be necessary to have both `feature_tables_conf` and `feature_tables_sources_conf`.

Reply:
Maybe we can tweak the format in a separate diff? Either way works, but above everything else i'd def prefer things to be consistent between this and @pyalex ingestion job.